### PR TITLE
feat: implement Julia parser

### PR DESCRIPTION
**Pull request by Octogonapus**
This PR implements parsing for Julia Manifests.

I'm adding this as part of https://github.com/aquasecurity/trivy/pull/4438

### CODE REVIEW

**Comment by Octogonapus** [comment]
@knqyf263 This is RTM from my end

**Comment by Octogonapus** [comment]
@knqyf263 Do I need to do anything else on this PR?

**Comment by Octogonapus** [comment]
@knqyf263 @simar7 Any chance this will ever get reviewed or merged?

**Comment by knqyf263** [comment]
@DmitriyLewen Can you please take a look first?

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse.go:52]: If i understand correctly - there are 2 manifest format for us:
- before 1.8
- after 1.8

But we don't use manifest format for parsing. So we can just compare with `nil` and remove `ManifestFormat`.
```suggestion
	if man.ManifestFormat == nil {
		man = Manifest{
			JuliaVersion:   "unknown",
			Dependencies:   oldDeps,
		}
	}
```

Correct me, if i have missed something.

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse.go:72]: If i understand correctly - we can use `uuid` for `ID` field - https://pkgdocs.julialang.org/v1/toml-files/#Multiple-packages-with-the-same-name

Can you also add test for case when there are 2 dependencies with same names, but different versions and uuid.

File: pkg/julia/manifest/parse_test.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse_test.go:63]: use already sorted Libs:
https://github.com/aquasecurity/go-dep-parser/blob/8dd9c9c78a18c7829657cc3c61c168981661fa16/pkg/julia/manifest/parse.go#L86-L87

File: pkg/julia/manifest/parse_test.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse_test.go:68]: same

File: pkg/julia/manifest/parse_test.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse_test.go:24]: You only use `Manifest.toml` files. 
Remove other files ( like `Project.toml`, `README.md`, etc...) files, please.

File: pkg/julia/manifest/parse_test.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse_test.go:59]:
```suggestion
			gotLibs, gotDeps, err := NewParser().Parse(f)
```

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse.go:92]: These names can confusing.
What if use something like this:
```suggestion
func parseDependencies(pkgId string, pkgDeps []string, allDeps map[string][]Dependency, juliaVersion string) *types.Dependency {
```

**Review by DmitriyLewen — COMMENTED** [review]
Hello @Octogonapus 
Thanks for your work!

I left some comments.
Take a look, please.

File: pkg/julia/manifest/parse.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse.go:69]: Added

File: pkg/julia/manifest/parse_test.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse_test.go:23]: Removed

File: pkg/julia/manifest/parse.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse.go]: Done

File: pkg/julia/manifest/parse.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse.go]: Done - required a small rewrite

**Comment by Octogonapus** [comment]
Thank you for looking at this. Review comments have been addressed.

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [reply pkg/julia/manifest/parse.go:52]:
```suggestion
		if _, err := r.Seek(0, io.SeekStart); err != nil {
		     return nil, nil, xerrors.Errorf("seek error: %w", err)
	}
```

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse.go:201]: I am not sure that we need both `PrimitiveDependency` and `DecodedDependency`.
What if we will add :
```
DependsOs    []string       `toml:"-"`       // list of dependent UUID's.
``` 
 field here - https://github.com/aquasecurity/go-dep-parser/blob/275db58029ca208af1928c0ab925929e37804444/pkg/julia/manifest/parse.go#L27-L31
and handle uuids for `["A", "B"]` case in this function.
Something like that:
```suggestion
// Decodes a primitive dependency using the metadata from parse time.
func decodeDependency(man *PrimitiveManifest, dep *PrimitiveDependency, metadata *toml.MetaData) (*DecodedDependency, error) {
	// Try to decode as []string first where the manifest looks like deps = ["A", "B"]
	var possibleDeps []string
	err := metadata.PrimitiveDecode(dep.Dependencies, &possibleDeps)
	if err == nil {
		var possibleUuids []string
		for _, depName := range possibleDeps {
			primDep := man.Dependencies[depName]
			if len(primDep) > 1 {
				// it is case for [deps.A.deps]
				// return error
			}
			possibleUuids = append(possibleUuids, primDep[0].UUID)
		}
		dep.DependsOs = possibleUuids
		return dep, nil
	}
```
`[deps.A.deps]` case already uses UUID

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse.go:85]: pkgID still contains version.

**Review by DmitriyLewen — COMMENTED** [review]
@Octogonapus left some idea. Can you take a look?

File: pkg/julia/manifest/parse.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse.go:203]: That definition of `PrimitiveDependency` can't get parsed: `toml: line 10 (last key "deps.A.deps"): incompatible types: TOML value has type map[string]interface {}; destination has type slice`. That error is why I delayed the parsing until later in `decodeDependency`.

File: pkg/julia/manifest/parse.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse.go]: Removed

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [reply pkg/julia/manifest/parse.go:201]: Perhaps i confused you.
I meant that we stay your solution for `PrimitiveDependency`, but add 1 more field - https://gist.github.com/DmitriyLewen/22431f167fdb110b5b17a90b6071051b#file-parse-go-L28-L33

Also we need to use `PrimitiveDecode` twice ( as you did that):
https://gist.github.com/DmitriyLewen/22431f167fdb110b5b17a90b6071051b#file-parse-go-L140-L172

My code in gist needs refactoring, but i hope you will understand my logic.
PS. sorry, i didn't pull your changes for ID.

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse.go:153]: If i remember correctly - `maps.Values` returns values in random order. Can you add sort here?

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse.go:142]: I think we can add sort UUIDs here

File: pkg/julia/manifest/parse.go
Comment (DmitriyLewen) [inline pkg/julia/manifest/parse.go:40]: Looks like we need to return `manMetadata` here too.

**Review by DmitriyLewen — COMMENTED** [review]
left 3 small comments

File: pkg/julia/manifest/parse.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse.go:153]: Done

File: pkg/julia/manifest/parse.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse.go:142]: Done

File: pkg/julia/manifest/parse.go
Comment (Octogonapus) [reply pkg/julia/manifest/parse.go:40]: No need, because the old deps format can always be parsed into that object. We only need the parse metadata for the new manifest format.

Comment (DmitriyLewen) [reply pkg/julia/manifest/parse.go:40]: okay, thanks for this explain!

**Review by DmitriyLewen — APPROVED** [review]
Thank you for this work!
Now this PR seems good to me.

@knqyf263 can you take a look and merge this PR, if it is okay for you?

**Comment by Octogonapus** [comment]
@DmitriyLewen @knqyf263 bump

**Comment by Octogonapus** [comment]
@knqyf263 @DmitriyLewen bump, what else is needed for merge?

**Comment by DmitriyLewen** [comment]
Hello @Octogonapus 
I don't have permission to merge this PR.

@knqyf263 has some priority tasks. 
Give us a little more time, please.

**Comment by knqyf263** [comment]
We've not seen any demand for [this feature](https://github.com/aquasecurity/trivy/discussions/4437) so far. Once this PR gets merged, maintainers need to keep maintaining. We probably want to suspend this task until the community needs it.

**Comment by Octogonapus** [comment]
@knqyf263 I understand. In the meantime, I asked the Julia community if others wanted this feature. They have added their support to https://github.com/aquasecurity/trivy/discussions/4437

**Comment by knqyf263** [comment]
>@knqyf263 I understand. In the meantime, I asked the Julia community if others wanted this feature. They have added their support to https://github.com/aquasecurity/trivy/discussions/4437

OK, so let's merge your PR! But as we're not so familiar with Julia, we probably need support from the community.

**Comment by DmitriyLewen** [comment]
Hello @Octogonapus 
Can you reopen and update you PR for Trivy now?

**Comment by Octogonapus** [comment]
Thank you!

> we probably need support from the community

Yes I can help maintain this.

> Can you reopen and update you PR for Trivy now?

Yes I will.

**Pull request merged**
